File: examples/source/binary_search/binary_search.cpp
#include <cstdint>

/// what if `arr === nullptr` or `target === nullptr`?
/// does alive2 know this?
int32_t binary_search(const int32_t *arr, int64_t size, const int32_t *target) {
    int32_t left = 0;
    int32_t right = size - 1;

    while (left <= right) {
        int32_t mid = left + ((right - left) / 2);
        
        if (arr[mid] == *target) {
            return mid;
        }
        
        if (arr[mid] < *target) {
            left = mid + 1;
        } else {
            if (mid == 0) {
                break;
            }
            right = mid - 1;
        }
    }
    
    return -1;
}

File: examples/source/binary_search/binary_search.rs
pub fn binary_search(arr: &[i32], target: &i32) -> i32 {
    let mut left: i32 = 0;
    let mut right: i32 = arr.len().wrapping_sub(1) as i32;

    while left <= right {
        let mid = left.wrapping_add((right.wrapping_sub(left)) / 2);
        let mid_idx = mid as usize;

        if arr[mid_idx] == *target {
            return mid;
        }

        if arr[mid_idx] < *target {
            left = mid.wrapping_add(1);
        } else {
            if mid == 0 {
                break;
            }
            right = mid.wrapping_sub(1);
        }
    }

    -1
}
